src/DataProtection/DataProtection/src/Cng/CbcAuthenticatedEncryptor.cs

```diff
@@ -56,222 +57,162 @@ public CbcAuthenticatedEncryptor(Secret keyDerivationKey, BCryptAlgorithmHandle
-    private byte[] DoCbcDecrypt(BCryptKeyHandle symmetricKeyHandle, byte* pbIV, byte* pbInput, uint cbInput)
-    {
-        // BCryptDecrypt mutates the provided IV; we need to clone it to prevent mutation of the original value
-        byte* pbClonedIV = stackalloc byte[checked((int)_symmetricAlgorithmBlockSizeInBytes)];
-        UnsafeBufferUtil.BlockCopy(from: pbIV, to: pbClonedIV, byteCount: _symmetricAlgorithmBlockSizeInBytes);
-
-        // First, figure out how large an output buffer we require.
-        // Ideally we'd be able to transform the last block ourselves and strip
-        // off the padding before creating the return value array, but we don't
-        // know the actual padding scheme being used under the covers (we can't
-        // assume PKCS#7). So unfortunately we're stuck with the temporary buffer.
-        // (Querying the output size won't mutate the IV.)
-        uint dwEstimatedDecryptedByteCount;
-        var ntstatus = UnsafeNativeMethods.BCryptDecrypt(
-            hKey: symmetricKeyHandle,
-            pbInput: pbInput,
-            cbInput: cbInput,
-            pPaddingInfo: null,
-            pbIV: pbClonedIV,
-            cbIV: _symmetricAlgorithmBlockSizeInBytes,
-            pbOutput: null,
-            cbOutput: 0,
-            pcbResult: out dwEstimatedDecryptedByteCount,
-            dwFlags: BCryptEncryptFlags.BCRYPT_BLOCK_PADDING);
-        UnsafeNativeMethods.ThrowExceptionForBCryptStatus(ntstatus);
-
-        var decryptedPayload = new byte[dwEstimatedDecryptedByteCount];
-        uint dwActualDecryptedByteCount;
-        fixed (byte* pbDecryptedPayload = decryptedPayload)
-        {
-            byte dummy;
-
-            // Perform the actual decryption.
-            ntstatus = UnsafeNativeMethods.BCryptDecrypt(
-                hKey: symmetricKeyHandle,
-                pbInput: pbInput,
-                cbInput: cbInput,
-                pPaddingInfo: null,
-                pbIV: pbClonedIV,
-                cbIV: _symmetricAlgorithmBlockSizeInBytes,
-                pbOutput: (pbDecryptedPayload != null) ? pbDecryptedPayload : &dummy, // CLR won't pin zero-length arrays
-                cbOutput: dwEstimatedDecryptedByteCount,
-                pcbResult: out dwActualDecryptedByteCount,
-                dwFlags: BCryptEncryptFlags.BCRYPT_BLOCK_PADDING);
-            UnsafeNativeMethods.ThrowExceptionForBCryptStatus(ntstatus);
+            refPooledBuffer.Dispose();
+            if (rentedBuffer is not null)
+            {
+                ArrayPool<byte>.Shared.Return(rentedBuffer, clearArray: true);
+            }
         }
 
-        // Decryption finished!
-        CryptoUtil.Assert(dwActualDecryptedByteCount <= dwEstimatedDecryptedByteCount, "dwActualDecryptedByteCount <= dwEstimatedDecryptedByteCount");
-        if (dwActualDecryptedByteCount == dwEstimatedDecryptedByteCount)
+#else
+        var pooledArrayBuffer = new PooledArrayBufferWriter<byte>(outputSize);
```

Comment:
Is there any reason we cannot use `using` for these?

```suggestion
        using var pooledArrayBuffer = new PooledArrayBufferWriter<byte>(outputSize);
```